src/sleep/sleep.c

```diff
@@ -40,6 +43,161 @@
 
 static char* arg_verb = NULL;
 
+static int read_fiemap(int fd, struct fiemap **ret) {
+        _cleanup_free_ struct fiemap *fiemap = NULL, *result_fiemap = NULL;
+        int extents_size;
+        struct stat statinfo;
+        uint32_t result_extents = 0;
+        uint64_t fiemap_start = 0, fiemap_length;
+        size_t fiemap_size = 0, result_fiemap_size = 0;
+
+        if (fstat(fd, &statinfo) != 0) {
+                log_debug_errno(errno, "Cannot determine file size (%s)\n %m",
+                                strerror(errno));
+                return -errno;
+        }
+        fiemap_length = statinfo.st_size;
+
+        fiemap = new(struct fiemap, 1);
+        if (!fiemap)
+                return log_oom();
+
+        result_fiemap = new(struct fiemap, 1);
+        if (!result_fiemap)
+                return log_oom();
+
+        /*  XFS filesystem has incorrect implementation of fiemap ioctl and
+         *  returns extents for only one block-group at a time, so we need
+         *  to handle it manually, starting the next fiemap call from the end
+         *  of the last extent
+         */
+        while (fiemap_start < fiemap_length) {
+                zero(*fiemap);
+
+                fiemap->fm_start = fiemap_start;
+                fiemap->fm_length = fiemap_length;
+                fiemap->fm_flags = FIEMAP_FLAG_SYNC;
+
+                /* Find out how many extents there are */
+                if (ioctl(fd, FS_IOC_FIEMAP, fiemap) < 0)
+                        return -errno;
+
+                /* Nothing to process */
+                if (fiemap->fm_mapped_extents == 0)
+                        break;
+
+                /* Result fiemap have to hold all the extents for the hole file */
+
+                /* Read in the extents */
+                extents_size = sizeof(struct fiemap_extent) *
+                                      (fiemap->fm_mapped_extents);
+
+                /* Resize fiemap to allow us to read in the extents */
+                if (!GREEDY_REALLOC(fiemap, fiemap_size,
+                                    sizeof(struct fiemap) + extents_size))
```

Comment:
`offsetof(struct fiemap, fm_extents) + extents_size`...

using offsetoff instead of sizeof() doesn't change much here on the compiler level, since the fiemap structure has no whitespace, but it's still nicer to read I think, and doesn't imply that the structure is set up in some specific way.